### PR TITLE
Loaded command stats

### PR DESCRIPTION
**Pull request by ErykS1**

### CODE REVIEW

File: src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Iterator;
+import java.util.Vector;
+
+public class CommandStatisticVisitor implements CommandVisitor {
+    private double totalLength = 0;
+    private Vector<Integer> previousCoordinates = null;
+    private Vector<Integer> currentCoordinates = null;
+    private int operateToCount = 0;
+
+    private int setPositionCount = 0;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java:12]: There is already CommandCounterVisitor which calculates those values (see PR #3 ). You should not duplicate this feature (you could extend it).

File: src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    @Override
+    public void visit(OperateToCommand operateToCommand) {
+        if (operateToCount == 0) {
+            operateToCount = 1;
+            setPositionCount = 0;
+        }
+        previousCoordinates = currentCoordinates;
+        currentCoordinates = new Vector<>();
+        currentCoordinates.add(operateToCommand.getX());
+        currentCoordinates.add(operateToCommand.getY());
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java:35]: awful waste of resources.

File: src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package edu.kis.powp.jobs2d.command;
+
+import java.util.Iterator;
+import java.util.Vector;
+
+public class CommandStatisticVisitor implements CommandVisitor {
+    private double totalLength = 0;
+    private Vector<Integer> previousCoordinates = null;
+    private Vector<Integer> currentCoordinates = null;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java:9]: Vector is not good abstraction here and null values could be avoided, e.g.:
https://docs.oracle.com/javase/8/docs/api/java/awt/Point.html

File: src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    private int operateToCount = 0;
+
+    private int setPositionCount = 0;
+
+    public double getTotalLength() {
+        return totalLength;
+    }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/CommandStatisticVisitor.java:16]: There should be at least two distances: the total distance of devices head and the distance of performed operateTo calls.

File: src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java [hunk 5]
```diff
@@ -70,6 +70,22 @@ public synchronized DriverCommand getCurrentCommand() {
...
+
+            int totalSubCommandsCount = statisticVisitor.getOperateToCount() + statisticVisitor.getSetPositionCount();
+            int operateToCommandsCount = statisticVisitor.getOperateToCount();
+            double totalCommandLength = statisticVisitor.getTotalLength();
+
+            return "Number of sub-commands used: " + totalSubCommandsCount + "\nNumber of operateTo commands used: " +
+                    operateToCommandsCount + "\nTotal command length: " + totalCommandLength;
+        } else {
+            return "No statistics to display";
+        }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java:86]: CommandManager is not the right place for preparing such string. This method should not dependent of "currentCommand". Place it somewhere else (maybe, separate class in command/visitor package?)

File: src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java [hunk 6]
```diff
@@ -2,11 +2,11 @@
...
 import java.util.Iterator;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
 
 import edu.kis.powp.jobs2d.Job2dDriver;
-import edu.kis.powp.jobs2d.command.CommandVisitor;
-import edu.kis.powp.jobs2d.command.DriverCommand;
-import edu.kis.powp.jobs2d.command.ICompoundCommand;
+import edu.kis.powp.jobs2d.command.*;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java:9]: let's minimize dependencies explicitly (avoiding '*' helps )

**Review by kowallus — CHANGES_REQUESTED** [review]
Different types of issues to fix.

**Comment by kowallus** [comment]
Remember to resolve conflicts!

File: src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java [hunk 7]
```diff
@@ -71,19 +73,7 @@ public synchronized DriverCommand getCurrentCommand() {
...
-            int operateToCommandsCount = statisticVisitor.getOperateToCount();
-            double totalCommandLength = statisticVisitor.getTotalLength();
-
-            return "Number of sub-commands used: " + totalSubCommandsCount + "\nNumber of operateTo commands used: " +
-                    operateToCommandsCount + "\nTotal command length: " + totalCommandLength;
-        } else {
-            return "No statistics to display";
-        }
+        return visitorStatisticsGenerator.generateStatistics(currentCommand);
     }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/manager/CommandManager.java:77]: OK, but this method is not necessary here. Clients can use VisitorStatisticsGenerator directly.

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        previousSetPosition = currentSetPosition;
+        currentSetPosition = true;
+        previousCoordinates.setLocation(currentCoordinates);
+        currentCoordinates.setLocation(setPositionCommand.getX(), setPositionCommand.getY());
+    }
+
+    @Override
+    public void visit(ICompoundCommand compoundCommand) {
+        totalLength = 0;
+        operateToLength = 0;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:48]: Here local variables should be used. If ICompoundCommand would contain another ICompoundCommand global fields would be overridden.
Fields totalLength & operateToLength should be set at the end of this visit method.

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+            DriverCommand command = iterator.next();
+            command.accept(this);
+            if (currentCoordinates != null && previousCoordinates != null) {
+                totalLength += Math.sqrt(Math.pow(currentCoordinates.getX() - previousCoordinates.getX(), 2) +
+                        Math.pow(currentCoordinates.getY() - previousCoordinates.getY(), 2));
+                if (!previousSetPosition && !currentSetPosition) {
+                    operateToLength += Math.sqrt(Math.pow(currentCoordinates.getX() - previousCoordinates.getX(), 2) +
+                            Math.pow(currentCoordinates.getY() - previousCoordinates.getY(), 2));
+                }
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:61]: OK, but I think that this logic could be implemented in operateTo visit. Nevertheless, I am not sure what would clearer.

**Review by kowallus — CHANGES_REQUESTED** [review]
Visitor implementation needs improvement. 
A deeply complex command example would be beneficial to validate the calculations

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        double operateToLength = 0;
+
+        Iterator<DriverCommand> iterator = compoundCommand.iterator();
+        while (iterator.hasNext()) {
+
+            DriverCommand command = iterator.next();
+            command.accept(this);
+            if (currentCoordinates != null && previousCoordinates != null) {
+                totalLength += Math.sqrt(Math.pow(currentCoordinates.getX() - previousCoordinates.getX(), 2) +
+                        Math.pow(currentCoordinates.getY() - previousCoordinates.getY(), 2));
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:63]: What about operateToLength? Where is it modified?

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+            }
+        }
+
+        this.totalLength += totalLength;
+        this.operateToLength += operateToLength;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:68]: I think operateToLength is always 0 :(.

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+            DriverCommand command = iterator.next();
+            command.accept(this);
+            if (currentCoordinates != null && previousCoordinates != null) {
+                totalLength += Math.sqrt(Math.pow(currentCoordinates.getX() - previousCoordinates.getX(), 2) +
+                        Math.pow(currentCoordinates.getY() - previousCoordinates.getY(), 2));
+            }
+        }
+
+        this.totalLength += totalLength;
```
Comment (kowallus) [inline src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:67]: Why += ? 
this.totalLength = totalLength; should work

**Review by kowallus — CHANGES_REQUESTED** [review]
I am not convinced about the visitor implementation

File: src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java (hunk 10 above)
Comment (JakubNowak22) [reply src/main/java/edu/kis/powp/jobs2d/command/visitor/CommandStatisticVisitor.java:63]: As requested in https://github.com/iis-powp-2024/powp_jobs2d_projects/pull/30#discussion_r1582757031 we moved that to the visitor receiving OperateToCommand.

**Pull request closed without merging**